### DISCUSSION TITLE
Airflow helm chart - Ingress 404 not found

### DISCUSSION THREAD

**Discussion by sararamasamy** [apache/airflow#26281] [General]
### Official Helm Chart version

1.6.0 (latest released)

### Apache Airflow version

airflow-1.6.0 

### Kubernetes Version

airflow-1.6.0 

### What happened

I am trying to implement ingress for the airflow and getting 404 not found. ( Its working through port-forward)

### What you think should happen instead

I was able to get it up using port-forward and not working through ingress

### How to reproduce

# Ingress configuration
ingress:
  # Enable all ingress resources (deprecated - use ingress.web.enabled and ingress.flower.enabled)
  enabled: true

  # Configs for the Ingress of the web Service
  web:
    # Enable web ingress resource
    enabled: true

    # Annotations for the web Ingress
    annotations: 
          kubernetes.io/ingress.class: nginx

    # The path for the web Ingress
    path: "/airflow"

    # The pathType for the above path (used only with Kubernetes v1.19 and above)
    #pathType: "ImplementationSpecific"

    # The hostname for the web Ingress (Deprecated - renamed to `ingress.web.hosts`)
    host: "quality-uat.tx.com"

    # The hostnames or hosts configuration for the web Ingress
    hosts: []
    # - name: ""
    #   # configs for web Ingress TLS
    #   tls:
    #     # Enable TLS termination for the web Ingress
    #     enabled: false
    #     # the name of a pre-created Secret containing a TLS private key and certificate
    #     secretName: ""

    # The Ingress Class for the web Ingress (used only with Kubernetes v1.19 and above)
    #ingressClassName: "nginx"

    # configs for web Ingress TLS (Deprecated - renamed to `ingress.web.hosts[*].tls`)
    tls:
      # Enable TLS termination for the web Ingress
      enabled: false
      # the name of a pre-created Secret containing a TLS private key and certificate
      secretName: ""

    # HTTP paths to add to the web Ingress before the default path
    precedingPaths: []

    # Http paths to add to the web Ingress after the default path
    succeedingPaths: []

  # Configs for the Ingress of the flower Service

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
I think that is not nearly enough to create a bug - this is at most a discussion. I think people here might help you to solve your problem as long as you show what kind of investigation you've done, what errors you observe and what kind of problems you see by following Kubernetes Ingress documentation (linked from here) https://airflow.apache.org/docs/helm-chart/stable/production-guide.html?highlight=ingress#ingress 

What are your finding by following the docs and applying them to the understanding of what is not working ?

**Reply by sararamasamy** [reply]
Here is the error: - - [09/Sep/2022:20:52:00 +0000] "GET /airflow/ HTTP/1.1" 404 493 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/105.0.0.0 Safari/537.36"

It was pretty straight forward one. Is there anyother setting I am missing for Ingress? Like base url in config?

  annotations:
    kubernetes.io/ingress.class: nginx
    meta.helm.sh/release-name: airflowv10
    meta.helm.sh/release-namespace: airflow

spec:
  rules:
    - host: quality-uat.tx.com
      http:
        paths:
          - path: /airflow
            pathType: Prefix
            backend:
              serviceName: airflowv10-webserver
              servicePort: airflow-ui

**Reply by potiuk** [reply]
No idea. What the K8S docs say ?

**Reply by potiuk** [reply]
Does the ingress run? Is it listening on the right port? do you have it correctly reflected in the service? Are the right services runnign? How do you get your connections in? 

Setting up ingress is a k8s "generic" knowledge - nothiing airflow specific so - you should set it up the same as all other ingresses you had - maybe you should see what are the differences vs. those seetings and understand where the problems are? 

What does your experience/check say ?

**Reply by sararamasamy** [reply]
I had set-up other applications which did work with the setting I have.  Thats why I am trying to understand, is it tested with this version of airflow or you have any references?

webserver:
8080 | airflow-ui | TCP

service:
spec:
  ports:
    - name: airflow-ui
      protocol: TCP
      port: 8080
      targetPort: 8080

Ingress:
HOST                         PATH     BACKEND                      PORT
quality-uat.tx.com | /airflow | airflowv10-webserver | airflow-ui

**Reply by potiuk** [reply]
Ingress should work. There might be a million reasons why it does not work for you, and you have to deal with the fact that without further investigation and time investment from your side anyone else might at most guess what is wrong. You hae much more power - you can actually debug and see what is wrong. You just need to exercise that power. So rather than asking someone else to remotely debugging your problem and possible configuration and resource issue, you should invest yourself into finding out what is wrong.

The only references you will find in the doc by searching for "ingress". But it if it does not, surely you need to debug it :). I've never set it up myself  - so hard for me to debug it for you, but I know plenty of people did.

But if I were you and if you have other ingresses setup, then I would recommend you to follow the standard ways how you debug helm charts (not only Airflow). It's a useful set of techniques and tools that you need if you are responsible for managing k8s instance that will help you to debug similar problems:

Helm has some fantastic tools for debugging where you can see what is produced by applying your values to helm templates https://helm.sh/docs/chart_template_guide/debugging/. By following it you will be able to see the exact resources helm produces and by following the K8S documentation you will be able to find out, what is wrong.

When you follow the https://airflow.apache.org/docs/helm-chart/stable/production-guide.html?highlight=ingress#ingress and link there where K8S ingresses are explained, and combine it with the regular tools that you can monitor and debug k8s deployment (I heartily recommend `k9s` - this way you can see what exactly resources are deployed, whether you have any error in resources and logs, you will likely be able to find a debug the problem further - this time with a running instance.

K8S is quite a  bit leaky abstraction - if things do not work, it's a deployement manager (your) job to understand it and debug some configuration issues. It's not a "black box"

Looking forward for the results of your investigation - it might be useful "from the trenches" example that others might find useful. 

And if you find an error in the helm chart, I heartily invite you to even contribute a solution back. Airflow is an open-source project which you get for free. and it is created by > 2100 people - mostly people like you who tried to use it, had some problems, fixed them - and contributed them back. This could be either a fix, or improvement, or clarification of documentation (maybe even an example/reference documentation on how to set up Ingress - that could be a fantastic contribution back in return for the free software you get),

**Comment by mconigliaro** [comment]
I'm having a similar problem, except the nginx ingress worked for me. it's the the ALB version that doesn't. Did anyone figure anything out? 

What I find odd is that when I `curl -v` the root URL, I get `Server: gunicorn` with `Location: /home`, so I'm sure my requests are making it to the app through the load balancer. But when I `curl -v` the same URL with `/home`, I get `Server: awselb/2.0` and `HTTP/1.1 404 Not Found`. So it's like the load balancer forwarding rules are messed up, and my ALB isn't forwarding `/home` requests to Airflow for some reason.

Here are the rules that were generated for my ALB.

```yaml
spec:
  ingressClassName: example-alb
  rules:
    - host: airflow.dev.example.com
      http:
        paths:
          - path: /
            pathType: ImplementationSpecific
            backend:
              service:
                name: airflow-webserver
                port:
                  name: airflow-ui
```

According to [the docs](https://kubernetes.io/docs/concepts/services-networking/ingress/#path-types):

> ImplementationSpecific` means "matching is up to the IngressClass. Implementations can treat this as a separate pathType or treat it identically to Prefix or Exact path types.

So I guess I'll have to find out what my `example-alb` ingressClass is doing...

**Reply by mconigliaro** [reply]
OK, I figured it out. `ImplementationSpecific` works like `Prefix` for my nginx ingressClass, but it works like `Exact` for my ALB ingressClass. That means only requests to `/` (exact match) were being forwarded to Airflow, and the rest were being rejected by the ALB with a 404. Setting `ingress.web.pathType = Prefix` in the helm chart was the fix for me.
